### ISSUE TITLE
Point/line support for OBJ

### ISSUE THREAD

**Issue by mosra** [ufbx/ufbx#56]
I remember you had to do something extra to support points and lines in FBX (as those are represented as degenerated triangles, IIRC), how hard would it be to extend this to OBJ? In particular, the following two files produce an empty scene and zero meshes, while I'd expect a point and a line mesh:

```obj
v 0.5 2 3
v 0 1.5 1
v 2 3 5.0

p 1
p 3
p 2
p 1
```

```obj
v 0.5 2 3
v 0 1.5 1
v 2 3 5.0

l 1 2
l 2 3
```

I'm mostly just striving for feature parity with Assimp so I can ditch it for OBJ processing, and this is the last missing bit. I don't need ufbx to support the advanced things like curves, just basic lines and points.

Thank you! :)

**Comment by bqqbarbhg** [comment]
Thanks for the issue! Good point, now with better degenerate triangle handling I could import such faces trivially by considering `p/l` in the same way as `f`, so that would require minimal code. I haven't done it yet as I wasn't sure if I should interpret lines as `ufbx_line_curve`, but those don't support materials or UVs, which can occur in .obj files. Also I don't really want to add a new element type for obj point clouds.

Configuration option in `ufbx_load_opts` is a possibility but I need to give this some thought, so I'll take a short while to mull over the design before I can fix this.

**Comment by bqqbarbhg** [comment]
Ah reminder for myself: `v/p` is *not* semantically equivalent to `f`. You can either have multiple points and a long line segment such as:

```
p 1 2 3 4 5
l 1 2 3 4 5
```

Actually that would give some merit to using `ufbx_line_curve` as then we could interpret:

```
l 1 2
l 2 3
```

as two line segments that happen to meet at vertex 2 and

```
l 1 2 3
```

as one three vertex line segment.

I prefer to do these decisions based on real files, @mosra do you happen to know ways of exporting these files from real modeling software that I could mess around with? (no worries if not, I'll look around :D)

**Comment by mosra** [comment]
Hmm, I don't think it makes sense to try to interpret consecutive line segment as a long line strip. From what I can gather, lines in an OBJ are just loose segments, possibly used to highlight wireframe or whatever other helper geometry. 

Just as a test I exported the default Blender cube with all faces deleted to an OBJ and got this -- there's not even any attempt to make those follow each other in any way:

```obj
# Blender 3.4.1
# www.blender.org
mtllib lines.mtl
o Cube
v 1.000000 1.000000 -1.000000
v 1.000000 -1.000000 -1.000000
v 1.000000 1.000000 1.000000
v 1.000000 -1.000000 1.000000
v -1.000000 1.000000 -1.000000
v -1.000000 -1.000000 -1.000000
v -1.000000 1.000000 1.000000
v -1.000000 -1.000000 1.000000
l 1 2
l 1 3
l 3 4
l 5 1
l 2 6
l 3 7
l 6 8
l 8 7
l 4 2
l 5 6
l 8 4
l 7 5
```

(Points got exported as just `v` entries, without any `p`'s, which is strange.)

And then, OBJ has _actual_ curve support (examples are at the bottom of http://paulbourke.net/dataformats/obj/), so if anything I would limit curve support to that feature only (and as I just saw in Blender, it has an option to export curves to OBJ as NURBs). But as I said above, I don't need those at the moment, just want to avoid losing information if the file contains `p` and `l` data.

- - - -

Completely tangential, but as a total coincidence I'm solving a possibly similar problem with wide line rendering. A cube could be rendered like this for example, with the top and bottom being a closed loop and the edges separate segments, or all as separate segments, or also just about any other way:

![image](https://user-images.githubusercontent.com/344828/213310882-5ce97d24-535b-4288-8b22-7b9490b380ab.png)

What I'm trying to show here is that, in my opinion, there's no point in trying to get some meaning from an unordered bag of line segments. If the lines are meant to be connected together, there always has to be some extra information specifying how, such as in a form of a curve, or by combining several "line strip" / "line loop" primitives together. And OBJ lines are too basic to contain such info.

**Comment by bqqbarbhg** [comment]
Thanks for the details!

I was leaning towards interpreting `p/l` as mesh geometry but this sealed the deal. I have actually been following http://paulbourke.net/dataformats/obj/ for the implementation (that's actually where the multi-point lines came from :D).

Bit of an aside but I'm unreasonably happy how easy the implementation turned out to be in [ufbx.c](https://github.com/ufbx/ufbx/pull/57/commits/4fb64855f915bd402860ac2f0529dfa20275e682#diff-4236ea6b21ebcb421db4ec6a3fa8bbe722f5b3fc92cfe3be5fe7ba0cb764c7e8), even including the multi-element line/point support.

**Comment by mosra** [comment]
Wow, that commit is all that was needed? Nice :)

Yesterday I commented with a tired brain, too many words to say too little, sorry. It should have been just a single sentence, "I don't have a clear opinion on how line strips could be handled, but I don't think they're important.". :D

**Comment by bqqbarbhg** [comment]
> Wow, that commit is all that was needed?

It broke some other stuff and I found a related bug with an empty face (line `f`) so a bit of cleanup, but for the `p/l` implementation, yes that was all that as needed!

And the comment was perfect for me, the example really helped convince me that `l` does not always semantically mean spline-type lines. Oh and also the .obj curve support is something I was looking into but there is a limit of how much rare *obj* complexity I want to include in a primarily fbx library :D

**Comment by mosra** [comment]
> there is a limit of how much rare obj complexity

Exactly, yes, it's important to make a cutoff point. If FBX itself has a corresponding curve-like feature and mapping OBJ to that feature is cheap code-wise, then it might make sense, but if it would require dedicated implementation that's just dead weight to 99% of users, then not.

**Comment by bqqbarbhg** [comment]
The most annoying part is that FBX does have a curve feature similar to *some* OBJ curves so it's a bit tempting to implement, but OBJ curves are a lot more flexible which might be confusing for users as some curve types work and others just get discarded. Currently I have enough testing that the importer doesn't crash or do anything else nasty when encountering the curves and that's probably enough for now :D

**Comment by bqqbarbhg** [comment]
Ditto as #55, should be fixed as of [v0.3.0](https://github.com/ufbx/ufbx/tree/v0.3.0), leaving the issue open as well in case for verification.

**Comment by mosra** [comment]
Can confirm it works now, thank you! :)
